Adding specification to handle the new C2x _BitInt type (#191)
ARM-software/abi-aa: design-documents/bit-precise-types.rst

..
   Copyright (c) 2023, Arm Limited and its affiliates.  All rights reserved.
   CC-BY-SA-4.0 AND Apache-Patent-License
   See LICENSE file for details

Rationale Document for ABI related to the C23 _BitInt type.
***********************************************************

Preamble
========

Background
----------

This document describes the rationale behind the ABI choices made for using the
bit-precise integral types defined in C2x.  These are ``_BitInt(N)`` and
``unsigned _BitInt(N)``.  These are defined for integral ``N`` and each ``N`` is
a different type.

The proposal for these types can be found in following link.
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2763.pdf

As the rationale mentioned, some applications have uses for a specific bit-width

> type.  In the case of writing C code which can be used to determine FPGA

Author: nitpick: /As the rationale *in that proposal* mentions/

hardware these specific bit-width types can lead to large performance and space
savings.

From the perspective of the Arm ABI we have some trade-offs to determine.  We
need to choose a representation for these objects in memory and in registers
along with the size and alignment of the objects.  The main trade-offs we have
identified in this case are on performance between different types of C-level
operations, whether certain hardware-level atomic operations are possible, and
general familiarity of programmers with the representation.

> 

Author: nitpick: maybe it reads more easily if this were a 3-item bullet list?

For this particular type we are estimating that the use of ``_BitInt`` types
will not be such that operations on these types are performance critical.

> 

Author: Less of a nitpick: I'm not sure we agreed that operations on these types
will never be performance critical, which the above sentence seems to
suggest?
I'd probably start by explaining the difficulty in defining ABI
rules/understanding the trade-offs as there is almost no code currently
using _BitInt, and it is next-to-impossible to predict exactly how it
will be used most often in the future.

Nonetheless, we need to make trade-off decisions.

Therefore, let's analyze some of the characteristics and features of
this type to try and make a best guess of how these types may get used
when target Arm CPUs.

(The specifics on not much code with _BitInt being performance critical
 could be repeated after the below analysis of use cases?)

Reviewer: The justifications in N2763 mention that `_BitInt` to some degree exists to allow code that is used in high-level synthesis to have better performance by not having to participate in the usual arithmetic promotions, like promoting up to `int` (which is about 32 bits), and only promoting up to the other `_BitInt` in size (so that adding `_BitInt(4)` and `_BitInt(7)` gives you a `_BitInt(7)`, not a 32-bit `int`, which is easier to translate to an FPGA).

At larger sizes, it's more of a nicety (being able to "natively" specify "bigints" that the compiler handles for you), but an extremely appealing nicety.

There seem to be two different regimes for these types.  The "small" regime
where bit-precise types could be stored in a single register, and the "large"

> regime where bit-precise types must span multiple registers.

Author: nitpick: maybe say general purpose register? Is the term "general
purpose register" used in other parts of the ABI docs?

Alignment and sizes
-------------------

These types must be at least byte-aligned so they are addressable, and at least
rounded to a byte boundary in size for ``sizeof``.  Since these types have an
aesthetic similarity to bit-fields one might expect better packing in an array
of ``_BitInt(24)`` than an array of ``int32_t`` types (i.e. packing as good as a
byte-array).  However, this would require a low alignment of such types and that
would mean loading and storing of even "small" sized ``_BitInt``'s crossing
cache boundaries -- leading to an unnecessary performance hit and hindering any
atomic operations on these.

> 

Author: nitpick on style: I'm wondering if it would be better to more explicitly
surface the pros and cons of a trade-off decision in the structure of
this document, rather than using full prose? For the example above, I'm
thinking of for example something like:

Trade-off decision 1: alignment of small-regime _BitInt

There are 2 obvious options:

option 1: 1 byte alignment
pros: list them
cons: list them

option 2: alignment rules same as 8/16/32/64-byte larger datatype (todo:
   maybe improve the wording of this title/option)
pros: list them
cons: list them

Discussion of option 1 vs option 2.
Describe why we chose option 2 based on the information and estimates we
currently have available.

(I noticed you have something similar further down for "Representation
in bits". It might be good to have the same structure for analyzing
different trade-off choices to be made in this document?)

of types for half of the values of ``N``, while choosing a double-register sized
chunk would allow atomic operations on types in the range between the register
and double-register sizes due to the associated extra alignment allowing
operations like ``CASP`` on aarch64 and ``LDRD`` on aarch32.  Moreover, the
majority of "large" size use-cases proposed so far are of power-of-two sizes
like sha256 which would not be in the range which suffers in space-cost from
this choice.  Finally, defining the ``_BitInt`` representation in this manner
means that on AArch32 a ``_BitInt(64)`` has the same alignment and size as a
``int64_t`` which is the largest size defined on that platform, and on AArch64
a ``_BitInt(128)`` has the same alignment and size as a ``__int128`` which is
the largest type defined on that platform.  This falls out of the fact that
double-register size maps to the largest integral Fundamental Data Type defined
on both platforms.

Hence for "large" sizes we are choosing to define a ``_BitInt(N)`` size and
alignment by treating them "as if" they are an array of double-register sized
Fundamental Data Types.

Representation in bits

> ----------------------

Author: Should the above title be "Representation in registers" instead of "...
in bits"?

- The AArch64 ``LDADD`` operations would not work naturally.

- Operations ``+,-,*,<<`` would all cause the need for masking at an ABI
  boundary.

- On AArch64 this would not match the expectation of developers, with
  ``_BitInt(8)`` not matching the representation of a ``char``.

Summary, suggestion, and reasoning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Overall it seems that for operations on small values option ``A`` is more
performant.  However, when acting on "large" values (i.e. greater than the size
of one register) it loses some of that benefit.  Storing to and from memory
would also come at a cost for this representation.  This is also likely to be
the most surprising representation for developers on an Arm platform.

Between option ``B`` and option ``C`` there is not a great difference in
performance characteristics.  However it should be noted that option ``C`` is
likely the most natural extension of the AArch32 PCS rules for unspecified bits
in a register containing a small Fundamental Data Type, while option ``B`` is
the most natural extension of the similar rules in AArch64 PCS.

As mentioned above, we do not expect operations on ``_BitInt`` types to be

> performance critical.  Given that providing a productive environment for

Author: Overall, I'm not sure if we can expect _BitInt to never be used in
performance-critical code.
If people will write code targeting FPGAs with _BitInt, presumably that
code is performance-critical (why would anyone otherwise target an FPGA
for that code?). If such code becomes popular, at some point, there may
be good reasons to use the same source code to target CPUs too. In that
case, the _BitInt code would be performance critical...

Reviewer: Perhaps it's more coherent to say "As mentioned above, we do not expect operations on ``_BitInt`` types _at large sizes_ to be performance critical." since the benefits of `_BitInt` allowing sidestepping the arithmetic promotions disappear for sizes larger than `_BitInt(32)`.

Reviewer: Perhaps it would be more coherent to say "As mentioned above, we do not expect operations on ``_BitInt`` types _with N greater than 32 bits_ to be performance critical." due to N2763's rationale.

..
   Copyright (c) 2023, Arm Limited and its affiliates.  All rights reserved.
   CC-BY-SA-4.0 AND Apache-Patent-License
   See LICENSE file for details

Rationale Document for ABI related to the C23 _BitInt type.
***********************************************************

Preamble
========

Background
----------

This document describes the rationale behind the ABI choices made for using the
bit-precise integral types defined in C2x.  These are ``_BitInt(N)`` and
``unsigned _BitInt(N)``.  These are defined for integral ``N`` and each ``N`` is
a different type.

> The proposal for these types can be found in following link.

Reviewer 2: following link:

> The proposal for these types can be found in following link.

Reviewer 2: ```suggestion
The proposal for these types can be found in following link:
```

https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2763.pdf

As the rationale in that proposal mentioned, some applications have uses for a
specific bit-width type.  In the case of writing C code which can be used to
determine FPGA hardware these specific bit-width types can lead to large
performance and space savings.

From the perspective of the Arm ABI we have some trade-offs and decisions to
make:

- We need to choose a representation for these objects in registers.
- We need to choose a representation, size and alignment of these objects in memory.

The main trade-offs we have identified in this case are:

- Performance of different C-level operations.
- Whether certain hardware-level atomic operations are possible.
- Size cost of storing values in memory.
- General familiarity of programmers with the representation.

> Since this is a new type there is large uncertainty on how it will be used by

Reviewer 2: new type,

> Since this is a new type there is large uncertainty on how it will be used by

Reviewer 2: ```suggestion
Since this is a new type, there is large uncertainty on how it will be used by
```

programmers in the future.  Decisions we make here may also influence future

> usage.  Nonetheless we must make trade-off decisions with this uncertainty.  The

Reviewer 2: Nonetheless -> Therefore,

> usage.  Nonetheless we must make trade-off decisions with this uncertainty.  The

Reviewer 2: ```suggestion
usage.  Therefore, we must make trade-off decisions with this uncertainty.  The
```

below attempts to analyze possible use-cases to make our best guess as to how
these types may be used when targeting Arm CPU's.

Use-cases known of so far
-------------------------

> There seem to be two different regimes for these types.  The "small" regime

Reviewer 2: ```suggestion
There are two different regimes for these types.  The "small" regime
```

Author: I adjusted to "We believe there are two different regimes" (point being that we could be mistaken and want to convey that this is how we view things rather than exactly what is happening).

where bit-precise types could be stored in a single general-purpose register,
and the "large" regime where bit-precise types must span multiple
general-purpose registers.

Here we discuss the use-cases for bit-precise integer types that we have
identified or been alerted to so far.

C code to describe FPGA behavior
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A major motivating use-case for this new type is to aid writing C code which
describes the desired behavior of an FPGA.  Without the availability of the new
``_BitInt`` type such C code would semantically have much wider types than
necessary when performing operations, especially given that operations on small
integral types promote their operands to ``int``.

If these wider than necessary operations end up in the FPGA they would use many
more many more logic gates than necessary.  Using ``_BitInt`` allows programmers
to write code which directly expresses what is needed.  This can ensure the FPGA
description generated saves space and has better performance.

The notable thing about this use-case is that though the C code may be run on an

> Arm architecture (e.g. for testing), the most critical use is when transferred

Reviewer 2: ```suggestion
Arm architecture (for example for testing), the most critical use is when transferred
```

Reviewer 2: Avoid using Latin contractions

Author: Removed the `e.g.` indirection entirely -- just said "may be run on an Arm architecture for testing".

> The notable thing about this use-case is that though the C code may be run on an
> Arm architecture (e.g. for testing), the most critical use is when transferred
> to an FPGA (i.e. not an Arm architecture).

Reviewer 2: ```suggestion
The notable thing about this use-case is that though the C code may be run on an
Arm architecture (for example. for testing), the most critical use is when transferred
to an FPGA (that is, not an Arm architecture).
```

> That said, if the operation that this FPGA performs becomes popular there may be

Reviewer 2: ```suggestion
However, if the operation that this FPGA performs becomes popular there may be
```

Author: Would like to not make this change -- I believe the change puts much more emphasis on the second point than it did before, and that's not emphasis that I believe should be there (at least that's what "However" sounds like to me).

Is this suggestion in order to make things easier to read for non-native speakers (as the suggestions to avoid latin contractions), or is it for some other reason?
(Would make the change if the point is ease-of-understanding for non-native speakers, reserve judgement for other motivations ;-).

> a need to run the code directly on CPU's in the future.

Reviewer 2: ```suggestion
a need to run the code directly on CPUs in the future.
```

> The requirements on Arm ABI's from this use-case are relatively small since the

Reviewer 2: ```suggestion
The requirements on the Arm ABI from this use-case are relatively small since the
```

type as opposed to a structure of three bytes for these data types.

If used for 24-bit color we believe that the performance of standard arithmetic
operations would not be critical.  This because each 24-bit pixel usually
represents three 8-bit colors so operations would unlikely be performed on the
single value as a whole.

We also believe that if used for 24-bit color it would be helpful to specify a
size and alignment scheme such that an array of ``_BitInt(24)`` is well packed.

Networking Protocols
~~~~~~~~~~~~~~~~~~~~

Many networking protocols have packed structures in order to minimize data sent
over the wire.  In order to be perfectly packed the code will need to use
bit-fields rather than bit-precise types for storage, since the bit-precise types
must be accessible and hence at least byte-aligned.

The incentives to use bit-precise integral types for networking code would be in
order to maintain the best representation of the operation that is being
performed.

One negative of using bit-precise integral types for networking code would be
that idioms like ``if (x + y > max_representable)`` where ``x`` and ``y`` have

> been loaded from small bit-fields would no longer be viable.  We have seen such

Reviewer 2: ```suggestion
been loaded from small bit-fields, would no longer be viable.  We have seen such
```

idioms for small values in networking code in the Linux kernel.  These are
intuitive to write but if ``x`` and ``y`` were to bit-precise types would not
work as expected.

If used in code handling networking protocols, our estimate is that the
arithmetic manipulation performed on such values will not be the main
performance bottleneck in networking protocols.  This estimation comes from the
belief that networking is often IO bound, and that small packed values in
networking protocols tend to have limited arithmetic performed on them.

Hence we believe that ease of debugging of values in registers may be more
critical than performance concerns in this use-case.

> To help the compiler optimize (e.g. for auto vectorization)

Reviewer 2: ```suggestion
To help the compiler optimize (for example, for auto vectorization)
```

Author: Adjusted this to `(possibly for auto vectorization)` -- hope that isn't falling into another ease-of-parsing pitfall ;-)

> To help the compiler optimize (e.g. for auto vectorization)

Reviewer 2: ```suggestion
To help the compiler optimize (for example, for auto vectorization)
```

guaranteed to use constant-time operations.

This use-case would only be using "large" bit-precise types.  Moreover all
relevant sizes are powers of two.

Translating some more esoteric languages to C
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

At the moment there exist some high-level languages which support arbitrary
bit-width integers.  Translating such languages to C would benefit from the new
C type.

We do not know of any specific use-case within these languages other than for
cryptography algorithms as above.  Hence the trade-offs in this space are
assumed to be based on the trade-offs from the cryptography use-case above.

We estimate the use of translating a more esoteric language to C to be less
common than writing code directly in C.  Hence the weighting of this use-case in
our trade-offs is correspondingly lower than others.

Possible transparent BigNum libraries
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

We have heard of interest in using the new bit-precise integer types to
implement transparent BigNum libraries in C.

Such a use-case unfortunately does not directly correspond to what kind of code

> will be using this (e.g. would this be algorithmic code or I/O bound code).

Reviewer 2: ```suggestion
will be using this (for example, would this be algorithmic code or I/O bound code).
```

> will be using this (e.g. would this be algorithmic code or I/O bound code).

Reviewer 2: ```suggestion
will be using this (for example, would this be algorithmic code or I/O bound code).
```

the unspecified bits above bit number 7 do not affect multiplication or division
operations.  Hence we do not look at variations of this kind.

For option ``A`` there is an extra choice around how "large" values are stored.
One could either have the "padding" bits in the least significant "chunk", or
the most significant "chunk".  Having these padding bits in the least
significant chunk would mean require something like a widening cast would
require updating every "chunk" in memory, hence we assume large values of option
``A`` would be represented with the padding bits in the most significant chunk.

Option ``A`` has the following benefits:

- For small values in memory, on AArch64, the operations like ``LDADD`` and
  ``LD{S,U}MAX`` both work (assuming the relevant register operand is
  appropriately shifted).

- Operations ``+,-,%,==,<=,>=,<,>,<<`` all work without any extra instructions
  (which is more of the common operations than other representations).

It has the following negatives:

- This would be a less familiar representation to programmers.  Especially the
  fact that a ``_BitInt(8)`` would not have the same representation in a

>   register as a ``char`` could cause confusion (e.g. when debugging, or writing

Reviewer 2: ```suggestion
  register as a ``char`` could cause confusion (For example, when debugging, or writing
```

Author: For some reason the "for example" reads clunkier than "e.g." to me.
I went with `cause confusion (we imagine when debugging or writing assembly code)`.
Hopefully that avoids the ease-of-parsing difficulty for non-native speakers.

other.

Alignment and sizes
-------------------

Options and their trade-offs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

These types must be at least byte-aligned so they are addressable, and at least
rounded to a byte boundary in size for ``sizeof``.

"Small" regime
//////////////
For the "small" regime there are 2 obvious options:

A. Byte alignment.
B. Alignment and size "as if" stored in the next-largest Fundamental Data Type.
   (Where the Fundamental Data Types are defined in the relevant PCS documents).

Option ``A`` has the following benefits:

- Better packing in an array of ``_BitInt(24)`` than an array of ``int32_t``.
  This is more relevant for bit-precise types than others since these types have
  an aesthetic similarity to bit-fields and hence programmers might expect good
  packing.

Option ``B`` has the following benefits (both following from the alignment being
greater than or equal to the size of the object in memory):

- Avoid a performance hit since loading and storing of these "small" sized
  ``_BitInt``'s will not cross cache boundaries.
- Atomic loads and stores can be made on these objects.
- The representation of bit-precise types of the same size as standard integer

>   types will have the same alignment and size in memory.

Reviewer 3: That last point is really the most important. The first 2 points are not too convincing since unaligned accesses are practically for free on modern cores, and LSE2 allows atomics on unaligned data.

Author: Ooh -- thanks!
I did not realise the cost of unaligned accesses were essentially free (useful to know).

Reviewer 3: Yes, optimization guides mention things like "The Neoverse N2 handles most unaligned accesses without performance penalties.", so you are unlikely to notice a penalty in typical compiled code.

Reviewer 4: It's not just about the atomic RMW operations, you also (I think) want ordinary reads and writes of _BitInt variables to be single-copy-atomic and implemtable with ordinary loads and stores. This implies they should be aligned so as not to cross cache boundaries, and also not packed along with other objects (i.e. no packing a _BitInt(24) together with a _BitInt(8) in a 32-bit word).

> 

Reviewer 3: Again, mentioning atomics as the first, most important thing does not make any sense. For large bitints you can't do atomics at all (no atomics on int arr[100] either!), small bitints up to int128 could support atomics similar to how bitfields could support atomics (typically using CAS due to needing extra normalization step). And it's not like LDSMAX/LDUMAX is a popular atomic given it's not even a supported builtin in GCC...

Author: Yeah -- I had not realised the order of bullet points would imply a priority order.
The atomics were not supposed to be presented as the most important reason -- will adjust.

To make sure I'm not missing something -- what extra normalization step are you referring to?
Is it the normalisation required for the set of operations which are not OK for a given representation (e.g. if choosing option `B` there would be a normalisation required for `/`, but not `+`) or is it something else you've noticed?

Reviewer 3: Yes, I mean the standard normalization. So while you can't use LDADD with option C, you can still use a CAS or load exclusive loop with an extra zero/sign-extend. Since you can do this for every atomic operation, being able to use LDADD or LDSMAX directly might be nice but shouldn't count as an advantage. Not being able to use atomics at all would be a disadvantage for a particular layout if atomic support is essential (and I'm not convinced it is for bitint).

- The AArch64 ``LD{S,U}MAX`` operations would not work naturally on small values
  of this representation.

- Operations ``/,%,==,<,>,<=,>=,>>`` and widening conversions on operands coming
  from an ABI boundary would require masking the operands.

- On AArch32 this could cause surprises to developers, given that on this
  architecture small Fundamental Data Types are have zero- or sign-extended
  extra bits.  So a ``char`` would not have the same representation as a
  ``_BitInt(8)`` on this architecture.

- If used in calls to variadic functions which were written for standard

>   integral types this can give surprising results.

Author: Hmm, honestly I'd not seriously considered this since the way that the PCS is defined is through a two step process of
"source code type" -> "fundamental data type"
then
"fundamental data type" -> "rule for PCS passing".

Now you've pointed it out it does seem reasonable.
Not sure whether the exceptional case could lead to confusion -- will invite other feedback to see what people think.
@smithp35 @stuij @rsandifo-arm

>   integral types this can give surprising results.

Author: Test @rearnsha (wanted to tag you, the comment box doesn't want to auto-complete your name, don't know if typing it fully out will properly link -- this is a test).

Summary, suggestion, and reasoning
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Overall it seems that option ``A`` is more performant for operations on small
values.  However, when acting on "large" values (i.e. greater than the size of
one register) it loses some of that benefit.  Storing to and from memory would
also come at a cost for this representation.  This is also likely to be the most
surprising representation for developers on an Arm platform.

> 

Reviewer 3: Note there is no reason why a compiler couldn't do register operations in format A, while passing arguments/storing as C and assuming B when loading bitints or using parameters/return values. A seems most useful as an optimized internal representation (which is entirely private) rather than as the storage format.

Author: :+1:  I'll add a paragraph explicitly mentioning that the compiler could always choose to use a different representation at non ABI boundaries.
